apps/emqx_bridge/src/emqx_bridge_lib.erl

```diff
@@ -66,36 +68,59 @@ downgrade_type(Type, Conf) when is_list(Type) ->
 %% A rule might be referencing an old version bridge type name
 %% i.e. 'kafka' instead of 'kafka_producer' so we need to try both
 external_ids(ConfRootKey, Type, Name) ->
-    case downgrade_type(Type, get_conf(ConfRootKey, Type, Name)) of
+    external_ids(?global_ns, ConfRootKey, Type, Name).
+
+external_ids(Namespace, ConfRootKey, Type, Name) ->
+    case downgrade_type(Type, get_conf(Namespace, ConfRootKey, Type, Name)) of
         Type ->
-            [external_id(Type, Name)];
+            [external_id(Namespace, ConfRootKey, Type, Name)];
         Type0 ->
-            [external_id(Type0, Name), external_id(Type, Name)]
+            [
+                external_id(Namespace, ConfRootKey, Type0, Name),
+                external_id(Namespace, ConfRootKey, Type, Name)
+            ]
     end.
 
 get_conf(BridgeType, BridgeName) ->
-    get_conf(undefined, BridgeType, BridgeName).
-get_conf(ConfRootKey, BridgeType, BridgeName) ->
+    get_conf(?global_ns, _ConfRootKey = undefined, BridgeType, BridgeName).
+
+get_conf(Namespace, ConfRootKey, BridgeType, BridgeName) ->
     case emqx_bridge_v2:is_bridge_v2_type(BridgeType) of
         true ->
             ConfRootKey1 =
                 case ConfRootKey of
                     undefined ->
-                        emqx_bridge_v2:get_conf_root_key_if_only_one(BridgeType, BridgeName);
+                        emqx_bridge_v2:get_conf_root_key_if_only_one(
+                            Namespace, BridgeType, BridgeName
+                        );
                     _ ->
                         ConfRootKey
                 end,
-            emqx_conf:get_raw([ConfRootKey1, BridgeType, BridgeName]);
+            get_raw_config(Namespace, [ConfRootKey1, BridgeType, BridgeName]);
         false ->
             undefined
     end.
 
 %% Creates the external id for the bridge_v2 that is used by the rule actions
 %% to refer to the bridge_v2
-external_id(BridgeType, BridgeName) ->
-    Name = bin(BridgeName),
-    Type = bin(BridgeType),
+external_id(Namespace, ConfRootKey, Type0, Name0) when is_binary(Namespace) ->
+    Kind =
+        case ConfRootKey of
+            actions -> <<"action">>;
+            sources -> <<"source">>
+        end,
+    Name = bin(Name0),
+    Type = bin(Type0),
+    <<"ns:", Namespace/binary, ":", Kind/binary, ":", Type/binary, ":", Name/binary>>;
```

Comment:
I had introduced this new notation to be explicit about namespace/kind for new namespaced rules, but I think it might be possible to get away with just `Type:Name`, since the rule will live inside a single namespace (or global) anyway.  Then it should search for actions/sources within its namespace.  :thinking:

Reply:
Not always. mqtt is both action and source.